### ISSUE TITLE
bufio: invalid use of UnreadByte

### ISSUE THREAD

**Issue by ramainen** [tamerh/xml-stream-parser#3]
go version go1.12.5 windows/amd64
xml-stream-parser: latest ("Skip XML declatarions at beginning" commit)
There is error while try to use Stream(),

for xml := range parserXML.Stream() {

xml.Err.Error() contains "Invalid xml"

I checked out, there is sendError() method shows this, and error actually contains "bufio: invalid use of UnreadByte".

He sterted at here: https://github.com/tamerh/xml-stream-parser/blob/master/xmlparser.go#L453

and before at here: 
https://github.com/tamerh/xml-stream-parser/blob/1b306d7894e38b5e7149f875d3617b6eea5487c5/xmlparser.go#L540

Next is bufio itself.

**Comment by ramainen** [comment]
It works well at another PC, go1.11.4 windows/amd64, if it helps

**Comment by ramainen** [comment]
in UnreadByte code:
// UnreadByte unreads the last byte. Only the most recently read byte can be unread.
//
// UnreadByte returns an error if the most recent method called on the
// Reader was not a read operation. Notably, Peek is not considered a
// read operation.

previous call is exactly Peek: https://github.com/tamerh/xml-stream-parser/blob/1b306d7894e38b5e7149f875d3617b6eea5487c5/xmlparser.go#L419

**Comment by ramainen** [comment]
If change code to:

		```
	//	b, err = x.reader.Peek(1) // this is needed because we cant unread 2 bytes consequtively
			z, err := x.readByte()

			if err != nil {
				return err
			}

			if z == '!' || z == '?' { // either comment or decleration
```

it works, but i do not know, how well.

**Comment by ramainen** [comment]
Can I make pull request with this?

**Comment by tamerh** [comment]
Hi @ramainen thanks for raising this issue.

I realised that i was using 1.11 and Peek function has changed in 1.12
https://github.com/golang/go/blob/release-branch.go1.11/src/bufio/bufio.go
https://github.com/golang/go/blob/release-branch.go1.12/src/bufio/bufio.go

unfortunately we can't replace the peek with the readByte like in your pull request, in that way we might lost bytes.

I am now reviewing and fixing the function that use Peek using 1.12. Let me know if you have suggestions.

**Comment by tamerh** [comment]
I fixed the function now seems that working with both 1.11 and 1.12

**Comment by tamerh** [comment]
Hopefully now fixed with parser version v1.14

**Comment by ramainen** [comment]
So fast, thank you
